### PR TITLE
(0.84.0) Fix `NetCDFOutputWriter` compression kwarg

### PR DESCRIPTION
**Pull request by Yixiao-Zhang**
To use compression in `NCDatasets` , the keyword for `defVar` should be `deflatelevel` instead of `compression`. `defVar` does not show any warning message for unrecognized keywords.

### CODE REVIEW

**Review by tomchor — APPROVED** [review]
Nice catch!
Do you have merge permission?

**Comment by navidcy** [comment]
I'm wondering if we should expend the remark on the `NetCDFOutputWriter` docstring regarding what `compression` is?

**Comment by glwagner** [comment]
Should we also use `deflatelevel` for the name of the `NetCDFOutputWriter` kwarg?

It'd be nice to have a test for this I think.

File: src/OutputWriters/netcdf_output_writer.jl
Comment (glwagner) [inline src/OutputWriters/netcdf_output_writer.jl:398]: What is the motivation for using the name `compression` in the user interface for `NetCDFOutputWriter`, but assigning it to `deflatelevel`? Is this standard / intuitive? Or should we also change `compression` to `deflatelevel`.

Comment (navidcy) [reply src/OutputWriters/netcdf_output_writer.jl:398]: I believe we should change compression to `deflatelevel` and in the docstring refer to NCDatasets.jl docs

Comment (navidcy) [reply src/OutputWriters/netcdf_output_writer.jl:398]: Although `compression` is more intuitive to `deflatelevel`....

Comment (tomchor) [reply src/OutputWriters/netcdf_output_writer.jl:398]: IMO compression is way more intuitive than `deflatelevel`, so I vote for leaving it the way it is. Does JLD2 have a similar option? What keyword do they use?

Comment (navidcy) [reply src/OutputWriters/netcdf_output_writer.jl:398]: I don't think the JLD2OutputWriter has this option, but looking at the JLD2 package I found:

https://juliaio.github.io/JLD2.jl/dev/compression/

We could use `compress::Boolean = true` as a keyword for both output writers and for the NetCDFOutputWriter also have a `deflatelevel` kwarg with some default value which will be passed if `compress = true`?

Comment (navidcy) [reply src/OutputWriters/netcdf_output_writer.jl:398]: Hm...

```
Note that CodecZlib.jl must be explicitly installed for compression/decompression to work. If the package is not already loaded JLD2 will try to dynamically load it, but it is recommended to do it explicitly, e.g. using JLD2, CodecZlib.
```

OK, so not sure if we want to deal with the JLD2OutputWriter in this PR, but in anticipation of potentially enabling compression for JLD2OutputWriter I still vote for a `compress::Bool` kwarg and then an additional kwarg with a default value, e.g., `deflatelevel::Int=4`.

Comment (tomchor) [reply src/OutputWriters/netcdf_output_writer.jl:398]: I'm okay with this solution

Comment (glwagner) [reply src/OutputWriters/netcdf_output_writer.jl:398]: I think we should pass kwargs to NCDatasets without changing their name.

Comment (navidcy) [reply src/OutputWriters/netcdf_output_writer.jl:398]: OK, this is a very good point!

Comment (navidcy) [inline src/OutputWriters/netcdf_output_writer.jl:398]:
```suggestion
                   deflatelevel=deflatelevel, attrib=default_dimension_attributes[dim_name])
```

File: src/OutputWriters/netcdf_output_writer.jl
Comment (navidcy) [inline src/OutputWriters/netcdf_output_writer.jl:576]:
```suggestion
               ("particle_id", "time"), deflatelevel=deflatelevel)
```

File: src/OutputWriters/netcdf_output_writer.jl
Comment (navidcy) [inline src/OutputWriters/netcdf_output_writer.jl:459]:
```suggestion
           deflatelevel=deflatelevel, attrib=output_attributes)
```

File: src/OutputWriters/netcdf_output_writer.jl
Comment (navidcy) [inline src/OutputWriters/netcdf_output_writer.jl:449]:
```suggestion
           deflatelevel=deflatelevel, attrib=output_attributes)
```

**Comment by navidcy** [comment]
@Yixiao-Zhang I made some suggestions but I couldn't make suggestions to everywhere so I also pushed some changes. Hope that this is alright!

**Review by navidcy — CHANGES_REQUESTED** [review]
As per the discussion in the PR, let's change the kwarg from `compression` to `deflatelevel`.

This is a breaking API change so @Yixiao-Zhang can you also bump up a minor release? That is, 
change the version number here:
https://github.com/Yixiao-Zhang/Oceananigans.jl/blob/36961523d3b35bfb08e9af779504b99beb60d585/Project.toml#L4
to 0.84.0 ?

**Comment by Yixiao-Zhang** [comment]
BTW, in the next release of `NCDatasets.jl`,`defVar` will throw an error if there is an unrecognized keyword \(See this [change](https://github.com/Alexander-Barth/NCDatasets.jl/issues/212)\).

That means that older versions of `Oceananigans.jl` \(\< 0.84\) will not work with newer versions of `NCDatasets.jl` \(\>0.12.17\).

Also for this reason, we want to release `Oceananigans.jl` v0.84 before the release of a newer version of  `NCDatasets.jl` \(v0.12.18 or 0.13.0\).

**Comment by navidcy** [comment]
> BTW, in the next release of `NCDatasets.jl`,`defVar` will throw an error if there is an unrecognized keyword (See this [change](https://github.com/Alexander-Barth/NCDatasets.jl/issues/212)).
> 
> That means that older versions of `Oceananigans.jl` (< 0.84) will not work with newer versions of `NCDatasets.jl` (>0.12.17).

Sounds that NCDatasets is gonna have a breaking change but _not_ a minor release?

**Comment by Yixiao-Zhang** [comment]
I do not mean that. I think it should be a major change.

When I wrote down that sentence, I did not realize that `compat` in `Project.toml` can prevent using newer versions of dependencies with breaking changes.

**Comment by glwagner** [comment]
> BTW, in the next release of `NCDatasets.jl`,`defVar` will throw an error if there is an unrecognized keyword (See this [change](https://github.com/Alexander-Barth/NCDatasets.jl/issues/212)).
> 
> That means that older versions of `Oceananigans.jl` (< 0.84) will not work with newer versions of `NCDatasets.jl` (>0.12.17).

Helpful to know that we will need that compat, thank you!

**Comment by glwagner** [comment]
> BTW, in the next release of `NCDatasets.jl`,`defVar` will throw an error if there is an unrecognized keyword (See this [change](https://github.com/Alexander-Barth/NCDatasets.jl/issues/212)).
> 
> That means that older versions of `Oceananigans.jl` (< 0.84) will not work with newer versions of `NCDatasets.jl` (>0.12.17).

That's good to know, thank you!

**Comment by glwagner** [comment]
> As per the discussion in the PR, let's change the kwarg from `compression` to `deflatelevel`.
> 
> This is a breaking API change so @Yixiao-Zhang can you also bump up a minor release? That is, change the version number here: https://github.com/Yixiao-Zhang/Oceananigans.jl/blob/36961523d3b35bfb08e9af779504b99beb60d585/Project.toml#L4 to 0.84.0 ?

Also the title of the PR should be prepended with the new version ie `(0.84.0) Fix NetCDF...`

**Review by navidcy — APPROVED** [review]

**Comment by navidcy** [comment]
OK, I'm merging this?

**Comment by navidcy** [comment]
@Yixiao-Zhang ready to merge?

File: src/OutputWriters/netcdf_output_writer.jl
Comment (navidcy) [inline src/OutputWriters/netcdf_output_writer.jl:213]:
```suggestion
                  and 9 means maximum compression). See [NCDatasets.jl documentation](https://alexander-barth.github.io/NCDatasets.jl/stable/variables/#Creating-a-variable).
```

File: src/OutputWriters/netcdf_output_writer.jl
Comment (navidcy) [inline src/OutputWriters/netcdf_output_writer.jl:213]:
```suggestion
                  and 9 means maximum compression). See [NCDatasets.jl documentation](https://alexander-barth.github.io/NCDatasets.jl/stable/variables/#Creating-a-variable)
```

**Comment by Yixiao-Zhang** [comment]
I think it is ready to merge.

**Comment by glwagner** [comment]
This PR: https://github.com/CliMA/Oceananigans.jl/pull/3091 also claims to be (0.84.0). Which is right? 

@tomchor I'm going to need help enforcing the new conventions that were discussed on #3147

**Comment by navidcy** [comment]
Both introduced breaking changes! And were merged the same day so we registered v0.84 after both were merged.

**Comment by navidcy** [comment]
I thought that it wouldn’t make sense to register v0.84 and 20min later v0.85….

**Comment by glwagner** [comment]
Ah ok

**Pull request merged**
